File: dsa-practice/arrays/codes/problemset3/prob3/prob3.cpp
#include <bits/stdc++.h>
using namespace std;

int kadane(int arr[], int n)
{
    int currSum = 0;
    int maxSum = INT_MIN;
    for (int i = 0; i < n; i++)
    {
        currSum += arr[i];
        if (currSum < 0)
        {
            currSum = 0;
        }
        maxSum = max(maxSum, currSum);
    }
    return maxSum;
}

int main()
{
    // int arr[7] = {4, -4, 6, -6, 10, -11, 12};
    int arr[4] = {1, -2, 3, -2};
    // TODO: get the total sum of the original array
    int totalSum = 0;

    // TODO: Find the nonWrap elements sum
    int nonWrapSum = kadane(arr,4);
    
    // TODO: Get the sum of non-contributing elements
    // 1- Reverse the signs of all element(positive to negative/negative to positive)
    for (int i = 0; i < 4; i++)
    {
        totalSum += arr[i];
        arr[i] = -(arr[i]);
    }

    // 2 - Apply kadane's algorithm to get the sum of non-contributing elements
    int nonContributingSum = kadane(arr, 4);

    // TODO:  Max Subarray sum = Total sum  of array - Sum of non-contributing elements
    int wrapSum = totalSum + nonContributingSum;
    cout << "MAX SUBARRAY SUM: " << max(wrapSum,nonWrapSum) << endl;
}

File: dsa-practice/arrays/codes/problemset3/prob3/prob3.js
/**
 * @param {number[]} nums
 * @return {number}
 */
function maxSubArraySum(a, size) {
  var max_so_far = Number.MIN_SAFE_INTEGER;
  var max_ending_here = 0;

  for (var i = 0; i < size; i++) {
    max_ending_here = max_ending_here + a[i];
    if (max_so_far < max_ending_here) max_so_far = max_ending_here;

    if (max_ending_here < 0) max_ending_here = 0;
  }
  return max_so_far;
}
const kadane = (nums) => {
  let currSum = 0;
  let maxSum = Number.MIN_SAFE_INTEGER;
  nums.forEach((num) => {
    currSum += num;
    // if(maxSum<currSum){
    //     maxSum = currSum;
    // }
    maxSum = Math.max(maxSum, currSum);
    if (currSum < 0) {
      currSum = 0;
    }
    
  });
  return maxSum;
};

var maxSubarraySumCircular = function (nums) {
  // TODO: get the total sum of the original array
  const totalSum = nums.reduce((a, b) => a + b);
  console.log("TOTAL SUM: ", totalSum);
  //  TODO: get the max subarray sum of nonWrapping elements
  const nonWrapSum = kadane(nums);
  if(nonWrapSum<0){
    return nonWrapSum;
  }
  console.log("nonWrapSum: ", nonWrapSum);
  // TODO: Get the sum of non-contributing elements
  const reverseSignArray = nums.map((num) => -num);
  // TODO:  Max Subarray sum = Total sum  of array - Sum of non-contributing elements
  const wrapSum = totalSum + kadane(reverseSignArray);
  console.log("wrapSum: ", wrapSum);
  return Math.max(nonWrapSum, wrapSum);
};
console.log(maxSubarraySumCircular([-3,-2,-3]))
console.log(kadane([-3, -2, -3]));
console.log(maxSubArraySum([-3, -2, -3],3))
